lib.importApply: init

roberth opened the pull request:
###### Description of changes

Brings variables from rich scopes to modules defined in separate files.

A helper for importing functions in files that return a module.

>    This solves the problem of being unable to pass a value from the lexical
    scope to a single module. This occurs when you return a module from a rich
    scope, such as that of a flake.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
lib.modules.importApply: init

Brings variables from rich scopes to modules defined in separate files.

A helper for functions in files that return a module.

infinisil commented:
I think it's an anti-pattern to make modules be functions that take parameters. You can generally always declare an option to do the same instead, which then comes with the advantage of merging, type checking, docs, etc.

roberth commented:
I'm trying to provide a better alternative to the pattern where people pass the flake `inputs` as a module argument. At that point, reuse is out of the window, so merging, checking and docs don't serve much of a purpose anymore.

I'd like to provide an alternative that is similarly convenient. We need a better pattern so that exposing a module doesn't just break by default and then require everyone to reinvent the wheel to figure out how it can be done properly.

Indeed splitting the module along one or more options instead of the lexical scope is another option, but it is a tough sell. It won't catch on as the default way of doing things. The unfortunate truth is that the blogosphere is bursting with the distrohopper archetype who just doesn't get around to exposing a module of their own, as they usually don't have a need to.
And this is how people learn. If you're doing any sort of development, these resources will be a net negative to you, as long as they keep making `inputs` a global variable.



> I think it's an anti-pattern to make modules be functions that take parameters.

I agree, but this function is only tangential to the anti-pattern. You *could* use `importApply` and the anti-pattern together, but I find this to be unlikely. For the intended purpose, the file looks just like a normal module, except with two function headers.

We can make a distinction between reasons for a function to exist. One is to bring some sort of "stable" things into scope, and the other is for invoking it multiple times for various circumstances. In other words, a distinction between dependency injection and general programming.

Indeed if general programming happens between *any* function header (including the normal module arguments) and the structures of the module system, such as the module attributes, `mkOption` arguments, etc, someone has engaged in what can almost be considered code generation - using the Nix language to generate code in the module system language.

Alternatively it can be seen as ad hoc extension of the module system. It is the module system's task to "compile" Nix to the module system, using well understood concepts. By programming your own extensions to the module system, with the explicit intent or by doing some ad hoc stuff in a module, you're adding unexpected and typically undocumented concepts to the module system, which makes it more complex and therefore worse. (And most of the time it shouldn't be necessary)

But this reasoning does not apply if it's _not_ general programming. Ignoring `with`, scope in Nix is as simple as it gets. Using a function for dependency injection is just that. It does not extend the module system with new concepts, and does not make it more complex.

I guess what you're going for is some sort of psychological argument, trying to somehow hide the capabilities of the Nix language, but I don't think that's very effective, and definitely not worth sacrificing composition for it (ie not providing a convenient alternative to global `inputs`).

Realistically there isn't much we can do about module system users resorting to general programming, `importApply` notwithstanding.

infinisil commented:
Oh I think I get it now, but I believe it can be explained with just some code:

This is okay:
```nix
let
  # Any variable in scope
  foo = "foo";
in
# Any module evaluation
evalModules {
  modules = [{
    imports = [
      # Pass the local `foo` variable into the module
      # Just as if the module was inlined
      (lib.importApply ./module.nix foo)
    ];
  }];
}
```

This is not okay:
```
let
  list = [ "foo" "bar" ];
in
# Any module evaluation
evalModules {
  modules = [{
    # Generate a module for each list element
    imports = map (lib.importApply ./module.nix) list
  }];
}
```

The former makes sense as seems justified as a use case, only the second is the anti-pattern. Did I get that right?

infinisil commented on lib/modules.nix:
```diff
@@ -1204,6 +1204,48 @@ let
+    The nix file must contain a function that returns a module. A module may
+    itself be a function. See the example below.
+
+    This function does not add support for deduplication and `disabledModules`,
+    although that could be achieved by wrapping the returned module. The reason
+    for this is that second argument of `importApply` may not be the same for
+    different invocations of `importApply`. Hence, the file name may not be a
+    key.
```
In which case I believe this should be changed. To support the use case of passing in variables from a scope, the deduplication will work just fine. Only the anti-pattern of generating modules would not work then (which is a good thing)

infinisil replied:
Yeah had the same thought. The distinguishing key can just be based on the import tree though, kind of what happens already with the anonymose keys, e.g. "key-0-2" would be the 2nd module in the imports list of the 0th `modules` argument.

roberth replied:
Relevant issue:
- https://github.com/NixOS/nixpkgs/issues/215496

sorki reviewed (approved):
I found this useful recently to make conditional module `imports` working as these cannot depend on module options itself. So :+1: from me

roberth commented:
@infinisil If we don't make `importApply` the norm, users will keep running into dependency injection problems.
- https://github.com/NixOS/nix/issues/11008

Note that it's not unique to flakes, but any framework that collects modules into attrsets for publishing.

aanderse commented:
hi there 👋 

maybe it's quite obvious to anyone but me... apologies if so... but could someone please explain to me what was so bad with the channels **syntax**? specifically the ability to reference a channel (analogous to a flake input) via `import <nixpkgs>` when `nixpkgs` was set in `NIX_PATH` environment variable.

let's forget implementation for a moment and pretend in a flakes world it is technically easy to implement functionally such that all my inputs are available via the old channel syntax (ie. `<flake_input_1>`, `<flake_input_2>`, etc...)  and still hermetic - why is this syntax bad/undesirable?

from my point of view it seems no worse than what is presented here and has the advantage that it is familiar syntax - but i assume i am missing something rather obvious on why what i am asking isn't good... any help? 

thanks!

roberth commented:
In a decentralized situation, where modules come from multiple repos, we'd have to assign more semantics to path lookups `<path>`.
Specifically, when referencing an input, which flake should it come from? Should it refer to the flake where the module is defined, or should it refer to the flake that imports it?

This is an instance of the main problem with path lookups, which is that they provide an extra, global context (global state if you will) that is hard to define and manipulate.

Meanwhile we have a lambda calculus inspired language, in which these things can already be expressed more explicitly and hermetically, without path lookups.
In other words, we replace an redundant, problematic, underdeveloped feature by reuse of the well developed core language.

It would perhaps be possible to develop the path lookup system to be usable with flakes somehow, but this has a significant opportunity cost, and will present a complex language feature that users will have to learn.
It would be far too risky to do this before developing and validating the required logic for this by using the language itself, which in my view we have hardly only started to do. If we get it wrong _again_, we are likely to be stuck with it, or half-stuck like flakes.

I guess fast adoption is a fun problem to have, but it also comes with a responsibility to handle significant features like this proactively and deliver an actual complete solution.
Speaking of flakes, a complete solution to the problem would also involve dependency version management built into flakes and/or the module system.

I guess it would help to talk about a specific example and how first class dependency management might look (ie using a library to implement it), but I feel like I'm already getting way off topic here.

aanderse commented:
> Specifically, when referencing an input, which flake should it come from? Should it refer to the flake where the module is defined, or should it refer to the flake that imports it?

definitely a good point and now that i think about it resolving this issue would either make rigid assumptions which go entirely against the spirit of this PR (and proper dependency management) **or** likely introduce a bunch of new syntax for specifying "channels" which sounds messy


> This is an instance of the main problem with path lookups, which is that they provide an extra, global context (global state if you will) that is hard to define and manipulate.
> 
> Meanwhile we have a lambda calculus inspired language, in which these things can already be expressed more explicitly and hermetically, without path lookups. In other words, we replace an redundant, problematic, underdeveloped feature by reuse of the well developed core language.

well put, point taken


> significant opportunity cost
> If we get it wrong again, we are likely to be stuck with it, or half-stuck like flakes.

indeed! and let me stress that i really appreciate the perspective you have brought to the [nix](https://github.com/NixOS/nix) project... i read every meeting update summary posted on discourse and i will take this opportunity to thank you for your contributions to that team, specifically in the form of advocating for what is needed to make this ecosystem thrive :heart:
consider my posts here an opportunity for me to learn, not an opportunity for me to make recommendations and tell you how important my less than informed opinions are :hand_over_mouth:


> Speaking of flakes, a complete solution to the problem would also involve dependency version management built into flakes and/or the module system.
> 
> I guess it would help to talk about a specific example and how first class dependency management might look (ie using a library to implement it), **but I feel like I'm already getting way off topic here.**

\> *emphasis added by me* <

off topic? yes and no. from a practical perspective, sure. by from a less practical perspective? you're right on topic, aren't you? you indirectly mention a more fundamental problem in this PR and suggest a solution which comes short of a real solution like "first class dependency management" which sounds absolutely lovely... but there are only so many hours in a day and sometimes we have to be pragmatic about solving problems... so i guess we're back to "off topic" :wink:

roberth commented:
Thank you for your kind words.

Who needs a topic ;D

I think ideally we'd have a better way of sharing this knowledge than issue and pr threads; maybe framed as design docs, and I wouldn't mind using a wiki for that. Not that I need it to be that way. You've just reminded me that this is valuable info, and it's not all that easy to find it here.

infinisil reviewed (approved):
Taking another think about this, I can see the need for this, so let's have this :D

infinisil commented on lib/modules.nix:
```diff
@@ -1204,6 +1204,48 @@ let
+        }
+
+  */
+  importApply =
```
Although hold on, not sure about the name, isn't there something more module system specific we could pick? Ideas:
- `importApplyModule`
- `moduleImportApply`
- `lib.modules.importApply`

roberth replied:
I've picked the latter, and edited the doc. Here's the relevant diff: https://github.com/NixOS/nixpkgs/compare/95e788d33b50aedbff158132157ed565b6b8c525..df67f1f

roberth committed:
lib.modules.importApply: Edit doc

Generally improve the quality. Notes:

- Not rendered to the manual yet, so probably the syntax could be
  improved, but I have no way to test this now.

- The docs use `arg` vs `staticArg` in the code. This is intentional,
  because the doc is pretty clear about the role of `arg` whereas
  the code exists in a context where ambiguities are more harmful.

roberth committed:
Format
